src/workerd/api/system-streams.c++

```diff
@@ -252,11 +252,99 @@ kj::Own<WritableStreamSink> newSystemStream(
   return kj::heap<EncodedAsyncOutputStream>(kj::mv(inner), encoding, context);
 }
 
+
+class WrappedAsyncIoStream final :
+    public ReadableStreamSource, public WritableStreamSink, public kj::Refcounted {
+  // A wrapper around a native `kj::AsyncIoStream` to enable a ReadableStream and WritableStream
+  // to be constructed from it. This stream does not support encoding.
+
+public:
+  explicit WrappedAsyncIoStream(kj::Own<kj::AsyncIoStream> inner, IoContext& context);
+  ~WrappedAsyncIoStream();
+
+  // ReadableStreamSource methods:
+  kj::Promise<size_t> tryRead(void* buffer, size_t minBytes, size_t maxBytes) override;
+
+  kj::Maybe<uint64_t> tryGetLength(StreamEncoding outEncoding) override;
+
+  kj::Maybe<Tee> tryTee(uint64_t limit) override;
+
+  // WritableStreamSink methods:
+  kj::Promise<void> write(const void* buffer, size_t size) override;
+  kj::Promise<void> write(kj::ArrayPtr<const kj::ArrayPtr<const kj::byte>> pieces) override;
+
+  kj::Maybe<kj::Promise<DeferredProxy<void>>> tryPumpFrom(
+      ReadableStreamSource& input, bool end) override;
+
+  kj::Promise<void> end() override;
+
+  void abort(kj::Exception reason) override;
+
+private:
+  kj::Own<kj::AsyncIoStream> inner;
+
+  IoContext& ioContext;
+};
+
+WrappedAsyncIoStream::WrappedAsyncIoStream(
+    kj::Own<kj::AsyncIoStream> inner, IoContext& context)
+    : inner(kj::mv(inner)), ioContext(context) { }
+
+WrappedAsyncIoStream::~WrappedAsyncIoStream() {
+  inner->shutdownWrite();
+}
+
+kj::Promise<size_t> WrappedAsyncIoStream::tryRead(
+    void* buffer, size_t minBytes, size_t maxBytes) {
+  return inner->tryRead(buffer, minBytes, maxBytes)
+      .attach(ioContext.registerPendingEvent());
+}
+
+kj::Maybe<uint64_t> WrappedAsyncIoStream::tryGetLength(StreamEncoding outEncoding) {
+  KJ_REQUIRE(outEncoding == StreamEncoding::IDENTITY);
```

Comment:
If the output encoding is not identity, I think we should just return nullptr. This function is saying, "I plan to read everything from this stream and immediately encode it in `outEncoding`. How many bytes is that going to be?"